lib/oxsdatatypes/src/parser.rs

```diff
@@ -125,412 +83,525 @@ impl From<DateTimeError> for XsdParseError {
-            opt(preceded(
-                char('.'),
-                take_while(|c: char| c.is_ascii_digit()),
-            )),
-        ))),
-        |v| {
-            let value = Decimal::from_str(v)?;
-            if Decimal::from(0) <= value && value < Decimal::from(60) {
-                Ok(value)
-            } else {
-                Err(XsdParseError {
-                    kind: XsdParseErrorKind::OutOfIntegerRange {
-                        value: value.as_i128() as u8,
-                        min: 0,
-                        max: 60,
-                    },
-                })
-            }
-        },
-    )(input)
+fn second_frag(input: &str) -> Result<(Decimal, &str), XsdParseError> {
+    let (number_str, input) = decimal_prefix(input);
+    let (before_dot_str, _) = number_str.split_once('.').unwrap_or((number_str, ""));
+    if before_dot_str.len() != 2 {
+        return Err(XsdParseError::msg(
+            "Seconds must be encoded with two digits",
+        ));
+    }
+    let number = Decimal::from_str(number_str)?;
+    if number < Decimal::from(0) || number >= Decimal::from(60) {
+        return Err(XsdParseError::msg("Seconds must be between 00 and 60"));
+    }
+    if number_str.ends_with('.') {
+        return Err(XsdParseError::msg(
+            "Seconds are not allowed to end with a dot",
+        ));
+    }
+    Ok((number, input))
+}
+
+// [63]   timezoneFrag ::= 'Z' | ('+' | '-') (('0' digit | '1' [0-3]) ':' minuteFrag | '14:00')
+fn timezone_frag(input: &str) -> Result<(TimezoneOffset, &str), XsdParseError> {
+    if let Some(left) = input.strip_prefix('Z') {
+        return Ok((TimezoneOffset::UTC, left));
+    }
+    let (sign, input) = if let Some(left) = input.strip_prefix('-') {
+        (-1, left)
+    } else if let Some(left) = input.strip_prefix('+') {
+        (1, left)
+    } else {
+        (1, input)
+    };
+
+    let (hour_str, input) = integer_prefix(input);
+    if hour_str.len() != 2 {
+        return Err(XsdParseError::msg(
+            "The timezone hours must be encoded with two digits",
+        ));
+    }
+    let hours = i16::from_str(hour_str)?;
+
+    let input = expect_char(
+        input,
+        ':',
+        "The timezone hours and minutes must be separated by ':'",
+    )?;
+    let (minutes, input) = minute_frag(input)?;
+
+    if hours > 13 && !(hours == 14 && minutes == 0) {
+        return Err(XsdParseError::msg(
+            "The timezone hours must be between 00 and 13",
+        ));
+    }
+
+    Ok((
+        TimezoneOffset::new(sign * (hours * 60 + i16::from(minutes)))?,
+        input,
+    ))
 }
 
-// [62]   endOfDayFrag ::= '24:00:00' ('.' '0'+)?
-fn end_of_day_frag(input: &str) -> XsdResult<'_, (u8, u8, Decimal)> {
-    map(
-        recognize(tuple((
-            tag("24:00:00"),
-            opt(preceded(char('.'), many1(char('0')))),
-        ))),
-        |_| (24, 0, 0.into()),
-    )(input)
+fn safe_parse<T>(
```

Comment:
The name seems more general than the functionality. Maybe something liek `ensure_complete`?